### PR TITLE
Allow merging dicts of different types

### PR DESCRIPTION
**Pull request by SebastianSpeitel**
Currently merging dicts via `|` is only allowed for mappings with the same keys and values.
But according to the PEP (584) any mapping is allowed.

### CODE REVIEW

**Comment by Akuli** [comment]
I'm not sure if we should do this. On the one hand, it works, but on the other hand, `foo | bar` should likely be an error when mixing the values of `foo` and `bar` isn't meaningful.

Sometimes mixed data types won't result in an error. For example, if you do `json.dumps(foo | bar)`, you won't know that something is wrong until you look at the resulting string.

**Comment by JelleZijlstra** [comment]
I think this makes sense for `__or__` and `__ror__`, but not for `__ior__`. We already do something similar for `set.__or__` (https://github.com/python/typeshed/blob/de0f94005f327cc1bfa45f12ab6ee3b4dc518eb8/stdlib/builtins.pyi#L858). It's true that this could hide some bugs as Akuli mentioned, but I can easily see people merging e.g. a `dict[str, str]` and a `dict[str, str | None]`.

We also do this for the i methods on set though, and we're missing the r methods there. Both of those seem like bugs.

**Comment by SebastianSpeitel** [comment]
> on the other hand, foo | bar should likely be an error when mixing the values of foo and bar isn't meaningful.

What do you mean by meaningful?

> For example, if you do `json.dumps(foo | bar)`, you won't know that something is wrong until you look at the resulting string of JSON.

Is that example a problem of the typing? Anything not taking an `Any` won't allow it.

> but I can easily see people merging e.g. a `dict[str, str]` and a `dict[str, str | None]`.

The example I was thinking of was people typing the key as a union of literals.

**Comment by SebastianSpeitel** [comment]
> I think this makes sense for `__or__` and `__ror__`, but not for `__ior__`.
That's right. I will change that. Should I amend or add a new commit?

**Comment by Akuli** [comment]
>What do you mean by meaningful?

I think `os.environ | {1: 2}` should be an error. I would expect a type checker to prevent mixing environment variables with integers.

**Comment by SebastianSpeitel** [comment]
> > What do you mean by meaningful?
> 
> I think `os.environ | {1: 2}` should be an error. I would expect a type checker to prevent mixing environment variables with integers.

But according to the PEP that would be allowed and works.

**Comment by JelleZijlstra** [comment]
> Should I amend or add a new commit

Always add a new commit. Force pushing makes review harder.

> I think os.environ | {1: 2} should be an error

But people will likely get an error anyway when they try to use their new `dict[int | str, int | str]`.

**Comment by SebastianSpeitel** [comment]
> > I think os.environ | {1: 2} should be an error
> 
> But people will likely get an error anyway when they try to use their new `dict[int | str, int | str]`.

That's what I thought. The code that uses the merged dict should have control over what it accepts.

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -826,8 +826,8 @@ class dict(MutableMapping[_KT, _VT], Generic[_KT, _VT]):
     __hash__: None  # type: ignore
     if sys.version_info >= (3, 9):
         def __class_getitem__(cls, item: Any) -> GenericAlias: ...
-        def __or__(self, __value: Mapping[_KT, _VT]) -> Dict[_KT, _VT]: ...
-        def __ror__(self, __value: Mapping[_KT, _VT]) -> Dict[_KT, _VT]: ...
+        def __or__(self, __value: Mapping[_T1, _T2]) -> Dict[Union[_KT, _T1], Union[_VT, _T2]]: ...
+        def __ror__(self, __value: Mapping[_T1, _T2]) -> Dict[Union[_KT, _T1], Union[_VT, _T2]]: ...
         def __ior__(self, __value: Mapping[_KT, _VT]) -> Dict[_KT, _VT]: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:831]:
```suggestion
        def __ior__(self, __value: Mapping[_KT, _VT]) -> Dict[_KT, _VT]: ...  # type: ignore
```
To make mypy not complain

**Review by srittau — APPROVED** [review]

**Pull request merged**